<a href="https://colab.research.google.com/github/epic-coder97/Data-602/blob/main/Homework_%7C_Week_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Homework | Week 4
## Nikita Dharmadhikari

*Reference : PPT and Notebooks*


## 1. Simple Regression Using statsmodel 


In [4]:
# importing Libraries
import pandas as pd
import numpy as np
import statsmodels.api as sm

''' Statsmodels is a great Python library for a lot of basic and inferential statistics. 
It also provides basic regression functions using an R-like syntax, so it's commonly used by statisticians. '''
# ols - stands for Ordinary least squares, we'll use this
from statsmodels.formula.api import ols


/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


- Import Boston datasets from scikit-learn datasets and load it 


In [5]:
from sklearn.datasets import load_boston

In [6]:
boston = load_boston()
print(boston.DESCR)


.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

### - Checking the Datatype
### - printing the value of the boston_dataset to understand what it contains

In [7]:
type(boston)
print(boston.keys())

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])


- data: contains the information for various houses
- target: prices of the house
- feature_names: names of the features
- DESCR: describes the dataset

<p>To know more about the features use boston_dataset.DESCR</p>

In [8]:
boston.DESCR

".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000

### Converting it into a Dataframe

In [9]:
bostondf = pd.DataFrame(boston.data, columns=boston.feature_names)
bostondf.head()

CRIM    ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31   0.0  0.538  ...  1.0  296.0     15.3  396.90   4.98
1  0.02731   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  396.90   9.14
2  0.02729   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  392.83   4.03
3  0.03237   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.63   2.94
4  0.06905   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  396.90   5.33

[5 rows x 13 columns]

### Describing the dataset

In [10]:
bostondf.describe()

CRIM          ZN       INDUS  ...     PTRATIO           B       LSTAT
count  506.000000  506.000000  506.000000  ...  506.000000  506.000000  506.000000
mean     3.613524   11.363636   11.136779  ...   18.455534  356.674032   12.653063
std      8.601545   23.322453    6.860353  ...    2.164946   91.294864    7.141062
min      0.006320    0.000000    0.460000  ...   12.600000    0.320000    1.730000
25%      0.082045    0.000000    5.190000  ...   17.400000  375.377500    6.950000
50%      0.256510    0.000000    9.690000  ...   19.050000  391.440000   11.360000
75%      3.677083   12.500000   18.100000  ...   20.200000  396.225000   16.955000
max     88.976200  100.000000   27.740000  ...   22.000000  396.900000   37.970000

[8 rows x 13 columns]

Define the data/predictors as the pre-set feature names

In [11]:
bostondf.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT'],
      dtype='object')

## Put the target (housing value -- MEDV) in another DataFrame

In [12]:
df = pd.DataFrame()

df['MEDV'] = boston.target
bostondf['MEDV'] = boston.target

In [13]:
df.columns

Index(['MEDV'], dtype='object')

## - Print first five rows 


In [14]:
df.head()

MEDV
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

## - Define 'RM' as X or the feature and 'MEDV' as y or the label

X : Feature or independent variable
Y : target or dependent variable

In [15]:
m = ols('MEDV ~ RM',bostondf).fit()
print(m.summary())

                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.484
Model:                            OLS   Adj. R-squared:                  0.483
Method:                 Least Squares   F-statistic:                     471.8
Date:                Sun, 27 Feb 2022   Prob (F-statistic):           2.49e-74
Time:                        17:26:58   Log-Likelihood:                -1673.1
No. Observations:                 506   AIC:                             3350.
Df Residuals:                     504   BIC:                             3359.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -34.6706      2.650    -13.084      0.0

- Skewness: not symmetric
- Kurtosis: tall
- Durbin-Watson: Positive autocorrelation




---
## 2. Multiple Regression Using statsmodel 
- Use 'RM' and 'LSTAT' as features and keep 'MEDV' as the target 



In [16]:
from sklearn.metrics import mean_squared_error, r2_score

In [26]:
X = bostondf[['RM', 'LSTAT']] 
y = bostondf['MEDV']
predictions = m.predict(X)
est = sm.OLS(y, X).fit()
print(est.summary())

                                 OLS Regression Results                                
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.948
Model:                            OLS   Adj. R-squared (uncentered):              0.948
Method:                 Least Squares   F-statistic:                              4637.
Date:                Sun, 27 Feb 2022   Prob (F-statistic):                        0.00
Time:                        18:22:28   Log-Likelihood:                         -1582.9
No. Observations:                 506   AIC:                                      3170.
Df Residuals:                     504   BIC:                                      3178.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

does a coefficient of determination of R2= .948 mean? 
<p>An R2 between 0 and 1 indicates the extent to which the dependent variable is predictable. In this case R2 of 0.948 means 94.8% of the variance Y (dependent variable) is predictable from X (independet variable).</p>

-Is there any autocorrelation? If not, why? 

<p>Answer : Durbin-Watson is a test of autocorrelation. Values from 0 to 2 mean positive autocorrelation .In this case the Durbin-Watson value is 0.834 which indicates that there is a POSTIVE AUTOCORRELATION.</p>

-Are the data coming from a normal distribution? Why? 

<p>Answer :Kurtosis: it is a measure of outliers in the distribution.
The normal distribution has a value of 3.
In this case Kurtosis value is greater than 3 which means 'tall distribution'.</p>

## 3. Multiple Regression Using scikitlearn 


In [18]:
from sklearn import linear_model
from sklearn import datasets 
data = datasets.load_boston()


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

In [20]:
# -Define the data/predictors as 'RM' and 'LSTAT' 
df1 = pd.DataFrame()
df1= pd.DataFrame(data.data, columns=data.feature_names)
target = pd.DataFrame(data.target, columns=["MEDV"])

feature_names=df1[['RM', 'LSTAT']]

# -Put the target (housing value -- MEDV) in another DataFrame 

df1 = pd.DataFrame(data.data, columns=data.feature_names)
target = pd.DataFrame(data.target, columns=["MEDV"])

In [27]:
# -Define the linear model as lm and fit

l = linear_model.LinearRegression()
model = l.fit(X,y)
predictions = l.predict(X)

#  -Print the predictions 

print("Predictions: \n", predictions)

Predictions: 
 [28.94101368 25.48420566 32.65907477 32.40652    31.63040699 28.05452701
 21.28707846 17.78559653  8.10469338 18.24650673 17.99496223 20.73221309
 18.5534842  23.64474107 23.10895823 22.9239452  24.65257604 19.73611045
 18.9297215  20.57377596 13.51732408 20.14832175 17.90896697 15.48764606
 18.35281036 16.56210901 18.74440281 18.34995811 23.51018847 24.94888935
 13.23095259 21.20092715 11.15596625 15.89983805 16.63398622 22.65107562
 21.07107521 22.81275431 22.53014238 29.46686594 33.15564849 30.0244275
 26.33937234 25.50630935 23.42747337 21.03183392 19.03080004 17.28696205
  6.35742724 16.77652446 20.38222834 23.73891662 28.42223975 23.78518476
 19.13293549 32.4841017  27.4553513  30.83048667 25.54262118 22.91599173
 19.44389291 19.76157796 27.21060683 26.99027936 29.66411644 27.68813019
 21.54751591 23.38578845 18.73350058 22.97822472 27.01833368 22.66525802
 25.99579831 25.61529631 26.24614271 24.92488095 22.94287168 23.32670532
 22.46574406 22.72305097 29.51629037 

In [28]:
# -Print the coefficient of determination 

l.score(X,y)

0.6385616062603403

In [29]:
# -Print coefficients 

l.coef_


array([ 5.09478798, -0.64235833])

In [30]:
# -Print intercept

l.intercept_


-1.358272811874489